### PR TITLE
Add support to run multiple H2 databases

### PR DESCRIPTION
**Pull request by marcellor**
Create H2 database schema when connection is over tcp. 
Add H2 schema to all SQL statements.

### CODE REVIEW

**Comment by capgen628** [comment]
Some test failures here. The test expectations need updating to include the specified schema name.

File: morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java [hunk 1]
```diff
@@ -438,7 +447,7 @@ protected String getSqlForYYYYMMDDToDate(Function function) {
   @Override
   protected String getSqlForDateToYyyymmdd(Function function) {
     String sqlExpression = getSqlFrom(function.getArguments().get(0));
-    return String.format("CAST(SUBSTRING(%1$s, 1, 4)||SUBSTRING(%1$s, 6, 2)||SUBSTRING(%1$s, 9, 2) AS DECIMAL(8))",sqlExpression);
+    return String.format("CAST(SUBSTRING(%1$s, 1, 4)||SUBSTRING(%1$s, 6, 2)||SUBSTRING(%1$s, 9, 2) AS INT)",sqlExpression);
```
Comment (jsimlo) [inline morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java]: Why changing to INT?

File: morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java [hunk 2]
```diff
@@ -449,7 +458,7 @@ protected String getSqlForDateToYyyymmdd(Function function) {
   protected String getSqlForDateToYyyymmddHHmmss(Function function) {
     String sqlExpression = getSqlFrom(function.getArguments().get(0));
     // Example for CURRENT_TIMESTAMP() -> 2015-06-23 11:25:08.11
-    return String.format("CAST(SUBSTRING(%1$s, 1, 4)||SUBSTRING(%1$s, 6, 2)||SUBSTRING(%1$s, 9, 2)||SUBSTRING(%1$s, 12, 2)||SUBSTRING(%1$s, 15, 2)||SUBSTRING(%1$s, 18, 2) AS DECIMAL(14))", sqlExpression);
+    return String.format("CAST(SUBSTRING(%1$s, 1, 4)||SUBSTRING(%1$s, 6, 2)||SUBSTRING(%1$s, 9, 2)||SUBSTRING(%1$s, 12, 2)||SUBSTRING(%1$s, 15, 2)||SUBSTRING(%1$s, 18, 2) AS BIGINT)", sqlExpression);
```
Comment (jsimlo) [inline morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java]: Why changing to BIGINT? Are DECIMALs slower or cause issues?

File: morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java [hunk 3]
```diff
@@ -567,77 +576,48 @@ protected String getSqlForAddMonths(Function function) {
...
+    sqlBuilder.append(Joiner.on(", ").join(intoFields));
+
+    // Add key fields
+    sqlBuilder.append(") KEY(");
+    Iterable<String> keyFields = Iterables.transform(statement.getTableUniqueKey(), new com.google.common.base.Function<AliasedField, String>() {
+      @Override
+      public String apply(AliasedField field) {
+        return field.getImpliedName();
+      }
+    });
```
Comment (wnicholson) [inline morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java]: `AliasedField::getImpliedName` is more succinct, I think.

File: morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java [hunk 4]
```diff
@@ -567,77 +576,48 @@ protected String getSqlForAddMonths(Function function) {
...
+    // Now add the into clause
+    sqlBuilder.append(schemaNamePrefix(statement.getTable()));
+    sqlBuilder.append(destinationTableName);
+    sqlBuilder.append("(");
+    Iterable<String> intoFields = Iterables.transform(statement.getSelectStatement().getFields(), new com.google.common.base.Function<AliasedField, String>() {
+      @Override
+      public String apply(AliasedField field) {
+        return field.getImpliedName();
+      }
+    });
```
Comment (wnicholson) [inline morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java]: AliasedField::getImpliedName?

File: morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java [hunk 5]
```diff
@@ -135,7 +134,17 @@ public H2Dialect() {
...
+
+    StringBuilder truncateStatement = new StringBuilder();
+
+    truncateStatement.append("TRUNCATE TABLE ");
+
+    truncateStatement.append(schemaNamePrefix(table))
+                     .append(table.getName());
+
+    statements.add(truncateStatement.toString());
+    return statements;
```
Comment (jsimlo) [inline morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java]: Avoid introducing new StringBuilders, only reuse them when they are there already.
Java compiler does StringBuilders already for us automatically, there is no added benefit.

https://niels.nu/blog/2016/stringbuilder.html Java is getting better :)

File: morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2.java [hunk 6]
```diff
@@ -62,6 +64,8 @@ public String formatJdbcUrl(JdbcUrlElements jdbcUrlElements) {
         .append(jdbcUrlElements.getPort() == 0 ? "" : ":" + jdbcUrlElements.getPort())
         .append("/mem:") // this means we're going to use a remote in-memory DB which isn't ideal
         .append(jdbcUrlElements.getDatabaseName());
+
+        builder.append(";INIT=CREATE SCHEMA IF NOT EXISTS " + firstNonNull(jdbcUrlElements.getSchemaName(), jdbcUrlElements.getDatabaseName()));
```
Comment (jsimlo) [inline morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2.java]: Shouldn't we remember which schema name have we created? SchemaName or DatabaseName?

File: morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java [hunk 7]
```diff
@@ -567,77 +576,48 @@ protected String getSqlForAddMonths(Function function) {
...
-   *     USING xmergesource
-   *     ON (Table.id = xmergesource.id)
-   *     WHEN MATCHED THEN UPDATE ...
-   *     WHEN NOT MATCHED THEN INSERT ...
-   * </pre>
-   *
    * @see org.alfasoftware.morf.jdbc.SqlDialect#getSqlFrom(org.alfasoftware.morf.sql.MergeStatement)
    */
   @Override
   protected String getSqlFrom(MergeStatement statement) {
```
Comment (jsimlo) [inline morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java]: This new Merge statement version does not work with statement.getIfUpdating() does it?
TestAccumulatingMergeStatement should fail if it does not

**Comment by marcellor** [comment]
Please ignore previous commits, I had to rebase the code.

File: morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2.java (hunk 6 above)
Comment (marcellor) [reply morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2.java]: When connecting to H2 over TCP we can either creates the schema explicitly or allow H2 to create it at init. We opted to the second approach.

File: morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2.java [hunk 8]
```diff
@@ -82,12 +86,7 @@ public String formatJdbcUrl(JdbcUrlElements jdbcUrlElements) {
     // The DEFAULT_LOCK_TIMEOUT=60000 sets the default lock timeout to 60
     //    seconds. When the value is not set, it takes default
     //    org.h2.engine.Constants.INITIAL_LOCK_TIMEOUT=2000 value
-    // The LOB_TIMEOUT defines how long a lob returned from a ResultSet is available post-commit, defaulting to 5 minutes (300000 ms)
-    //    Set this to 2 seconds to allow certain tests using lob fields to work
-    // The MV_STORE is a flag that governs whether to use the new storage engine (defaulting to true as of H2 version 1.4, false in prior versions)
-    //  Note that implementations of H2 prior to version 1.4.199 had an MVCC parameter used to allow higher concurrency.
-    //  This configuration has been removed and the old "PageStore" implementation (MV_STORE=FALSE) is no longer supported.
-    builder.append(";DB_CLOSE_DELAY=-1;DEFAULT_LOCK_TIMEOUT=60000;LOB_TIMEOUT=2000;MV_STORE=TRUE");
+    builder.append(";DB_CLOSE_DELAY=-1;DEFAULT_LOCK_TIMEOUT=60000");
```
Comment (okeating) [inline morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2.java]: Is there a reason for removing the LOB_TIMEOUT ? It has caused problems in the past particularly for integration tests that reset the start position frequently with a large number of business rules leading to a large memory footprint

Comment (jsimlo) [reply morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2.java]: There has been some rebasing, I assume https://github.com/alfasoftware/morf/pull/98/files is now the best summary of the actual changes going in..

**Pull request merged**